src/TwigComponent/src/Twig/TwigPreLexer.php

```diff
@@ -67,24 +70,54 @@ public function preLexComponents(string $input): string
                 }
             }
 
+            if ($this->consume('{% embed')) {
+                $inTwigEmbed = true;
+                $output .= '{% embed';
+                $output .= $this->consumeUntil('%}');
+
+                continue;
+            }
+
+            if ($this->consume('{% endembed %}')) {
+                $inTwigEmbed = false;
+                $output .= '{% endembed %}';
+
+                continue;
+            }
+
             $isTwigHtmlOpening = $this->consume('<twig:');
             $isTraditionalBlockOpening = false;
+
             if ($isTwigHtmlOpening || (0 !== \count($this->currentComponents) && $isTraditionalBlockOpening = $this->consume('{% block'))) {
                 $componentName = $isTraditionalBlockOpening ? 'block' : $this->consumeComponentName();
 
                 if ('block' === $componentName) {
                     // if we're already inside the "default" block, let's close it
-                    if (!empty($this->currentComponents) && $this->currentComponents[\count($this->currentComponents) - 1]['hasDefaultBlock']) {
+                    if (!empty($this->currentComponents) && $this->currentComponents[\count($this->currentComponents) - 1]['hasDefaultBlock'] && !$inTwigEmbed) {
                         $output .= '{% endblock %}';
 
                         $this->currentComponents[\count($this->currentComponents) - 1]['hasDefaultBlock'] = false;
                     }
 
                     if ($isTraditionalBlockOpening) {
+                        $prevPosition = $this->position;
+
                         // add what we've consumed so far
                         $output .= '{% block';
                         $output .= $this->consumeUntil('%}');
-                        $output .= $this->consumeUntilEndBlock();
+
+                        // If the last consumed string does not match the Twig's block name regex, we assume the block is self-closing
+                        $isBlockSelfClosing = '' !== preg_replace(
+                            Lexer::REGEX_NAME,
+                            '',
+                            trim(mb_substr($this->input, $prevPosition, $this->position - $prevPosition)))
+                        ;
```

Comment:
Instead of keeping calculating the content between the positions, can we use the return value of `$this->consumeUntil('%}');`?

Also, in the test, you have

```twig
{% block my_embed_block_1 "foo" %}
```

I can understand that, after the `preg_replace()`, the `""` would remain. And so `isBlockSelfClosing` would be true. But isn't this a valid syntax?

```twig
{% block my_embed_block_1 fooVariable %}
```

In this case, everything would be replaced, the result would be an empty string, and `isBlockSelfClosing` would be false, right? Could we, instead, (A) grab the content between `{% block` and `%}`, `trim()` the whitespace, then set `isBlockSelfClosing` to true if there are ANY non-whitespace characters left?

Reply:
> But isn't this a valid syntax?
> 
> ```
> {% block my_embed_block_1 fooVariable %}
> ```

Indeed.  (https://twig.symfony.com/doc/3.x/tags/extends.html#block-shortcuts)

```twig
{% block title page_title|title %}
```




(but pretty sure this is not the [hardest problem](https://twig.symfony.com/doc/3.x/tags/use.html) we may _theoretically_ have while mixing twig/html  :)

Reply:
> Instead of keeping calculating the content between the positions, can we use the return value of `$this->consumeUntil('%}');`?

Hum you're right, I will give it a try.


> Also, in the test, you have
> 
> ```
> {% block my_embed_block_1 "foo" %}
> ```
> 
> I can understand that, after the `preg_replace()`, the `""` would remain. And so `isBlockSelfClosing` would be true. But isn't this a valid syntax?

This is not true, when running the regex [`/[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*/A`](https://github.com/twigphp/Twig/blob/3.x/src/Lexer.php#L46) on `my_embed_block "foo"`, only the `my_embed_block` will be removed from the string, ` "foo"` will still be there.

See https://regex101.com/r/LVIP6p/1.

> ```
> {% block my_embed_block_1 fooVariable %}
> ```
> 
> In this case, everything would be replaced, the result would be an empty string, and `isBlockSelfClosing` would be false, right? Could we, instead, (A) grab the content between `{% block` and `%}`, `trim()` the whitespace, then set `isBlockSelfClosing` to true if there are ANY non-whitespace characters left?

Nope, ` fooVariable` will still be there after `preg_replace`, see https://regex101.com/r/FU3VmO/1.